core/modules/loader/WorkerServer.cc

```diff
@@ -0,0 +1,180 @@
+        sendData.reset(); // never send a response back for one of these, infinite loop.
+        break;
+    case LoaderMsg::MAST_WORKER_INFO:
+        _centralWorker->workerInfoReceive(data);
+        break;
+    case LoaderMsg::KEY_INSERT_REQ:
+        _centralWorker->workerKeyInsertReq(inMsg, data);
+        break;
+    case LoaderMsg::KEY_INFO_REQ:
+        _centralWorker->workerKeyInfoReq(inMsg, data);
+        break;
+    case LoaderMsg::WORKER_KEYS_INFO_REQ:
+        _centralWorker->workerWorkerKeysInfoReq(inMsg, data);
+        break;
+    case LoaderMsg::WORKER_RIGHT_NEIGHBOR:
+        _centralWorker->workerWorkerSetRightNeighbor(inMsg, data);
+        break;
+    case LoaderMsg::WORKER_LEFT_NEIGHBOR:
+        _centralWorker->workerWorkerSetLeftNeighbor(inMsg, data);
+        break;
+
+    // following not expected by worker
+    case LoaderMsg::KEY_INFO:
+    case LoaderMsg::MAST_INFO_REQ:
+    case LoaderMsg::MAST_WORKER_LIST_REQ:
+    case LoaderMsg::MAST_WORKER_INFO_REQ:
+    case LoaderMsg::MAST_WORKER_ADD_REQ:
+        // TODO add response for known but unexpected message.
+        sendData = replyMsgReceived(senderEndpoint, inMsg, LoaderMsg::STATUS_PARSE_ERR, "unexpected Msg Kind");
+    default:
+        sendData = replyMsgReceived(senderEndpoint, inMsg, LoaderMsg::STATUS_PARSE_ERR, "unknownMsgKind");
+    }
+
+    return sendData;
+}
+
+
+BufferUdp::Ptr WorkerServer::replyMsgReceived(boost::asio::ip::udp::endpoint const& senderEndpoint,
+                                              LoaderMsg const& inMsg,
+                                              int status, std::string const& msgTxt) {
+
+    if (status != LoaderMsg::STATUS_SUCCESS) {
+        LOGS(_log,LOG_LVL_WARN, "Error response Original from " << senderEndpoint <<
+                " msg=" << msgTxt << " inMsg=" << inMsg.getStringVal());
+    }
+
+    LoaderMsg outMsg(LoaderMsg::MSG_RECEIVED, inMsg.msgId->element, getOurHostName(), getOurPort());
+
+    // create the proto buffer
+    proto::LdrMsgReceived protoBuf;
+    protoBuf.set_originalid(inMsg.msgId->element);
+    protoBuf.set_originalkind(inMsg.msgKind->element);
+    protoBuf.set_status(LoaderMsg::STATUS_PARSE_ERR);
+    protoBuf.set_errmsg(msgTxt);
+    protoBuf.set_dataentries(0);
+
+    StringElement respBuf;
+    protoBuf.SerializeToString(&(respBuf.element));
+
+    auto sendData = std::make_shared<BufferUdp>(1000); // this message should be fairly small.
+    outMsg.appendToData(*sendData);
+    respBuf.appendToData(*sendData);
+    return sendData;
+}
+
+
+void WorkerServer::_msgRecieved(LoaderMsg const& inMsg, BufferUdp::Ptr const& data,
+                                boost::asio::ip::udp::endpoint const& senderEndpoint) {
+    bool success = true;
+    // This is only really expected for parsing errors. Most responses to
+    // requests come in as normal messages.
+    StringElement::Ptr seData = std::dynamic_pointer_cast<StringElement>(MsgElement::retrieve(*data));
```

Comment:
I'v seen this pattern many times, and it worries me. It's related to an earlier made (big) comment on a design of class StringRange.